src/main/java/fi/fmi/avi/model/taf/immutable/TAFImpl.java

```diff
@@ -178,6 +181,257 @@ public Builder withCompleteForecastTimes(final ZonedDateTime reference) {
             return this;
         }
 
+        public Builder withAllTimesComplete(final ZonedDateTime reference) {
+            requireNonNull(reference, "reference");
+            withCompleteIssueTimeNear(reference);
+            return withCompleteForecastTimes(getIssueTime()//
+                    .flatMap(PartialOrCompleteTimeInstant::getCompleteTime)//
+                    .orElse(reference));
+        }
+
+        /**
+         * Sets the TAF-specific message status.
+         * <p>
+         * Note, this method is provided for backward compatibility with previous versions of the API. In addition to
+         * setting the <code>status</code> value, it also sets other property values with the following logic:
+         * <ul>
+         *     <li>status is {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#CANCELLATION}: <code>reportStatus =</code>
+         *     {@link fi.fmi.avi.model.AviationWeatherMessage.ReportStatus#NORMAL}, <code>cancelMessage = true</code>, <code>missingMessage = false</code></li>
+         *     <li>status is {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#MISSING}: <code>reportStatus =</code>
+         *     {@link fi.fmi.avi.model.AviationWeatherMessage.ReportStatus#NORMAL}, <code>cancelMessage = false</code>, <code>missingMessage = true</code></li>
+         *     <li>status is {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#NORMAL}: <code>reportStatus =</code>
+         *     {@link fi.fmi.avi.model.AviationWeatherMessage.ReportStatus#NORMAL}, <code>cancelMessage = false</code>, <code>missingMessage = false</code></li>
+         *     <li>status is {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#AMENDMENT}: <code>reportStatus =</code>
+         *     {@link fi.fmi.avi.model.AviationWeatherMessage.ReportStatus#AMENDMENT}, <code>cancelMessage = false</code>, <code>missingMessage = false</code></li>
+         *     <li>status is {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#CORRECTION}: <code>reportStatus =</code>
+         *     {@link fi.fmi.avi.model.AviationWeatherMessage.ReportStatus#CORRECTION}, <code>cancelMessage = false</code>, <code>missingMessage = false</code></li>
+         * </ul>
+         *
+         * @param status the status to set
+         * @return builder
+         * @deprecated migrate to using a combination of {@link #setReportStatus(ReportStatus)}, {@link #setCancelMessage(boolean)} and
+         * {@link #setMissingMessage(boolean)} instead
+         */
+        @Override
+        @Deprecated
+        public Builder setStatus(final TAFStatus status) {
+            super.setStatus(status);
+            if (TAFStatus.CANCELLATION.equals(status)) {
+                super.setReportStatus(ReportStatus.NORMAL).setCancelMessage(true).setMissingMessage(false);
```

Comment:
This is already handled in my [next comment](https://github.com/fmidev/fmi-avi-messageconverter/pull/68#discussion_r523933303).

Reply:
So a cancel report is always considered as an amendment?

Reply:
The way I read the issue https://github.com/wmo-im/iwxxm/issues/125 was that a message *can* be both a cancellation and an amendment at the same time, but making it a cancellation would not always make it an amendment too. So currently setting a TAF as a cancelMessage sets it's ReportStatus to NORMAL instead of AMENDMENT. If this is incorrect, I'll correct it.

Reply:
As of my understanding, and according to our aviation meteorologists, yes.

Reply:
For any TAF a normal message must exist in order to cancel it. Only the first submission of a TAF message is considered normal, all following submissions are either amendments or corrections. Maybe in a rare case TAF cancellation status could be correction (I'm not sure, cannot find any mention on this), but in general it is always an _amendment_.

With SIGMETs and AIRMETs it is a bit different (which wmo-im/iwxxm#125 also concerns). Those are never amended, but a new normal message is sent to cancel a former one.

The EU [Acceptable Means of Compliance (AMC) and Guidance Material (GM) to Part-MET](https://www.easa.europa.eu/document-library/acceptable-means-of-compliance-and-guidance-materials/amc-and-gm-part-met) explicitly states how SIGMETs and AIRMETs shall be cancelled (without the `AMD` keyword), but provides only one example of a TAF cancellation, containing the `AMD` keyword.

Reply:
OK, this is now fixed in https://github.com/fmidev/fmi-avi-messageconverter/pull/68/commits/f8ec30544b6f942b71266b181661c303ad481676

The setCancelMessage does not affect the reportStatus, but calling setStatus(TAFStatus.CANCELLATION) sets the reportStatus to ReportStatus.AMENDMENT only in the case when the reportStatus is either not present or is set to NORMAL